lld/ELF/SyntheticSections.cpp

```diff
@@ -2711,6 +2714,581 @@ static uint32_t computeGdbHash(StringRef s) {
+        } else {
+          Abbrev *abbrev2 =
+              new (abbrevAlloc.Allocate()) Abbrev(std::move(abbrev));
+          abbrevSet.InsertNode(abbrev2, insertPos);
+          abbrevTable.push_back(abbrev2);
+          newCode = abbrevTable.size();
+          abbrev2->code = newCode;
+        }
+        chunk.nameIndices[i].abbrevCodeMap[oldAbbrev.Code] = newCode;
+      }
+    }
+  }
+
+  // Compute the merged abbrev table.
+  raw_svector_ostream os(abbrevTableBuf);
+  for (Abbrev *abbrev : abbrevTable) {
+    encodeULEB128(abbrev->code, os);
+    encodeULEB128(abbrev->tag, os);
+    for (DWARFDebugNames::AttributeEncoding a : abbrev->attributes) {
+      encodeULEB128(a.Index, os);
+      encodeULEB128(a.Form, os);
+    }
+    os.write("\0", 2); // attribute specification end
+  }
+  os.write(0); // abbrev table end
+  hdr.AbbrevTableSize = abbrevTableBuf.size();
+}
+
+void DebugNamesBaseSection::Abbrev::Profile(FoldingSetNodeID &id) const {
+  id.AddInteger(tag);
+  for (const DWARFDebugNames::AttributeEncoding &attr : attributes) {
+    id.AddInteger(attr.Index);
+    id.AddInteger(attr.Form);
+  }
+}
+
+std::pair<uint32_t, uint32_t> DebugNamesBaseSection::computeEntryPool(
+    MutableArrayRef<InputChunk> inputChunks) {
+  TimeTraceScope timeScope("Merge .debug_names", "entry pool");
+  // Collect the compilation units for each unique name. Speed it up using
+  // multi-threading as the number of symbols can be in the order of millions.
+  constexpr size_t numShards = 32;
+  const size_t concurrency =
+      bit_floor(std::min<size_t>(config->threadCount, numShards));
+  const size_t shift = 32 - countr_zero(numShards);
+  const uint8_t cuAttrSize = getMergedCuCountForm(hdr.CompUnitCount).first;
+  DenseMap<CachedHashStringRef, size_t> maps[numShards];
+  SmallVector<NameTableEntry, 0> nameVecs[numShards];
+  parallelFor(0, concurrency, [&](size_t threadId) {
+    for (auto i : seq(numChunks)) {
+      InputChunk &chunk = inputChunks[i];
+      for (NameIndex &ni : chunk.nameIndices) {
+        for (NameTableEntry &nte : ni.nameEntries) {
+          auto shardId = nte.hashValue >> shift;
+          if ((shardId & (concurrency - 1)) != threadId)
+            continue;
+
+          nte.chunkIdx = i;
+          for (IndexEntry &ie : nte.entries()) {
+            ie.abbrevCode = ni.abbrevCodeMap[ie.abbrevCode];
+            // Update the DW_IDX_compile_unit attribute (the last one ensured by
+            // `DebugNamesBaseSection::parse`).
+            auto &back = ie.attrValues.back();
+            back.attrValue += chunk.baseCuIdx;
+            back.attrSize = cuAttrSize;
+          }
+
+          auto &nameVec = nameVecs[shardId];
+          auto [it, inserted] = maps[shardId].try_emplace(
+              CachedHashStringRef(nte.name, nte.hashValue), nameVec.size());
+          if (inserted)
+            nameVec.push_back(std::move(nte));
+          else
+            nameVec[it->second].indexEntries.append(nte.indexEntries);
```

Comment:
This should be a std::move, to make sure we don't accidentally copy/change any of the IndexEntry addresses.

Reply:
The code is fine. 

There is no rvalue reference overload for `append` and such a overload would not make a difference as we would have to copy the elements one by one.